src/transformers/cache_utils.py

```diff
@@ -752,6 +752,15 @@ def update(
             self.key_cache[layer_idx] = torch.cat([self.key_cache[layer_idx], key_states], dim=-2)
             self.value_cache[layer_idx] = torch.cat([self.value_cache[layer_idx], value_states], dim=-2)
 
+        elif key_states.shape[-2] >= self.window_length - self.num_sink_tokens:
+            # Renew cache but keep sink tokens
+            sink_keys = self.key_cache[layer_idx][:, :, : self.num_sink_tokens]
+            remain_length = self.window_length - self.num_sink_tokens
+            self.key_cache[layer_idx] = torch.cat([sink_keys, key_states[:, :, -remain_length:]], dim=-2)
+
+            sink_values = self.value_cache[layer_idx][:, :, : self.num_sink_tokens]
+            self.value_cache[layer_idx] = torch.cat([sink_values, value_states[:, :, -remain_length:]], dim=-2)
+
```

Comment:
I *think* the values in the `remain_length` length will have to be rerotated, as in the `else` below 🤔 

For instance, in a sink cache with 1 sink token and a sliding window of 3, only position_ids in `[0, 1, 2, 3]` will be valid. If we pass an input with length 10:
- token with index `0` will be kept as sink as is
- token with index `1-6` will be discarded
- tokens with index `7-9` will have to be rerotated to indexes `1-3`

Another required change: in the case the cache is empty and the initial prompt is bigger than the cache, the current code (which simply appends) will also be incorrect :(

Reply:
Ah, this was done considering that user will call cache only from generate(). But before that, seems like I interpreted the implementation differently from you.

From what I see in code, if sink=1 and window=3, then our cache can't go beyond positions `[0, 1, 2]`. The existing code for update is 
```
self.key_cache[layer_idx][
                :, :, -self.window_length + self.num_sink_tokens + key_states.shape[-2] :
            ]
```
So, the keys are cropped to leave any number of older tokens minus sink, minus `new_keys_length`. Correct me if I'm wrong, in that case the current PR has to be reworked


If my understanding of code is correct and people use SinkCache with `generate()`, then this should work. Because we cropped inputs to the `window_length` (3 tokens) in modeling, but `(3 >= 3 - 1)`. The new 3 tokens already have position ids `[0,1,2]` so we can plug them in without rerotations. 

But I can add a rerotation in case people use SinkCache without generate and forget to crop inputs to the cache max length.

Reply:
> in the case the cache is empty and the initial prompt is bigger than the cache, the current code (which simply appends) will also be incorrect :(

Ah, yeah, that is another case to be fixed. I will also make a better doc on SinkCache in this PR or another with general "How to Use Cache" docs. Hope that will solve most of our issues from users

Reply:
Oops, you're right, the cache length is `self.window_length` (and not `self.window_length + self.num_sink_tokens`), the maximum number of positions in the example above is `3` indeed!
______________________________________

My comments are also referring to `generate`! 

I think I got what you meant -- in the model-side changes, you crop `input_ids` as `input_ids[:, -max_cache_length:]`, so `position_ids` will indeed be aligned by default. No need for rerotation. 

However, I think we're not handling the sink tokens correctly in a corner case :) Consider the following:
1. consider `window_length=256` and `num_sink_tokens=4`;
2. 1st generation with `max_new_tokens=1` and prompt of 2 tokens. These 2 tokens will go into the sink tokens;
3. 2nd generation with `max_new_tokens=1` and a prompt of 500 tokens. 

In this example, according to the current code, we will only have 2 sink tokens -- we should update sink tokens, but we don't. Moreover, because of the model-side changes, we won't be able to access the new sink tokens (they are discarded in `prepare_inputs_for_generation`).

As such:
1. We can't crop `input_ids`, as we need both ends of this tensor in some corner cases 😢 We **could** keep both ends of `input_ids`, discarding the middle, but that would be model-side code, which is undesirable. We should push all logic to the cache class.
2. Because of 1., we will need to rerotate

Reply:
Okey, I experimented with SinkCache a bit. My decision to make it work is to:

1. First of all, we ask users to not pass any input that is longer than window length to the `generate` with SinkCache.  Thus they have to crop if themselves when the length goes beyond and ofc they can't start off with a very long prompt. Why we can't keep inputs long until cache logic starts?

- In that case we need to keep position ids the same length (which forces us to keep attn mask also), otherwise RoPE will fail with shape mismatch.
- We can't do rerotation on the very long key which is beyond window length, because in that case we have to cache `cos` and `sin` indefinitely long inside SinkCache, We can;t know for sure how long the keys will be, and we need their original cosines/sines to rerotate back.
- Yes, in that case we might lose sink tokens, if they weren't filled in the first iteration. But afaik sink should be a very small number (2 to 4), and it's very unprobable that anyone will have less than 4 tokens in the first turn. We can better raise a warning in this case telling user that SinkTokens weren't filled in the first turn so we are filling it with the tokens we have at hand.

2. Secondly, I modified `cache_position` in `generate()` if we see that the input is already out of `max_cache_length` bounds. Following on (1), we need to adjust cache-position
3. Inside SinkCache, if we get more tokens than we can hold in general, we renew cache. In this place we add sink tokens if it wasn't filled and raise a warning

Reply:
Regarding the tests, the first one is passing on my end so I'm wondering if it's hardware related. Did it fail on T4 runners?

The second one is partially fixed by this PR. Partially because apparently we don't get the exact same match after `rerotations`, and it works if we crop inputs by leaving  only unprocessed tokens. I did a small test with rerotations and couldn't verify that `rerotate(rotate(key)) == key`. Can you plz look at this, I might be missing out on something?

Reply:
Update regarding rerotations, I inspected the original repo for SinkTokens and they never did something like that, neither they evicted tokens while generating. Their main point was about evicting cache in between turns, so that the new turn starts with 'sink + window_length - max_new_tokens' tokens. 

So I am wondering why we do it this way, as it degrades when window is very small compared to the max new tokens, even without iterative generation. For ex, if max_new_tokens=256 and window=30 tokens, the generation near the end is garbage

Reply:
@zucchini-nlp Many questions, many answers :D I've sorted them by my perceived importance, and added a number for follow-up comments/questions!

> (1) We can't do rerotation on the very long key which is beyond window length

We can also throw an exception in this case -- it is an edge case that's expensive to maintain, and there are alternatives (chunk the prompt and then feed it to the model). Personally, given how problematic `SinkCache` is, I think it is the best long-term solution for `transformers` :)

> (2) So I am wondering why we do it this way, as it degrades when window is very small compared to the max new tokens, even without iterative generation.

Nice dive here! In `transformers`, we cache the rotated KV tensors. There were three options:
1. Change what we cache
2. Cache both rotated and unrotated tensors
3. Re-rotate

Option 3. was the only viable choice. Rerotating is mathematically equivalent, but we also know that floating point operations can be problematic! Have you tried upcasting rerotations to FP32 (or even to FP64)? Maybe it can solve the numerical precision issue :D

> (3) I did a small test with rerotations and couldn't verify that rerotate(rotate(key)) == key. Can you plz look at this, I might be missing out on something?

See (2)

> (4) Regarding the tests, the first one is passing on my end so I'm wondering if it's hardware related. Did it fail on T4 runners?

The bulletproof check is to checkout to the commit that introduced `SinkCache`, where all tests were passing, and confirm that they are passing on your test hardware, too! If yes, then it means there are more regressions to fix :(

> (5) Why we can't keep inputs long until cache logic starts?

I believe this is the case at the moment, with `cache_positions` 🤔 (it may have not been in the past, before the transition to `cache_positions`, can't fully remember)